README.md
## neuralnexus - An open source, low-code machine learning framework built using Python.

CHANGELOG.txt
Change Log 
===========

0.0.1 (16/03/2023)
-------------------
- Ranking algorithm added

test.py
# Importing the Packages:
import optuna
import pandas as pd
from sklearn import linear_model
from sklearn import ensemble
from sklearn import datasets
from sklearn import model_selection

#Grabbing a sklearn Classification dataset:
X,y = datasets.load_breast_cancer(return_X_y=True, as_frame=True)

#Step 1. Define an objective function to be maximized.
def objective(trial):

    classifier_name = trial.suggest_categorical("classifier", ["LogReg", "RandomForest"])
    
    # Step 2. Setup values for the hyperparameters:
    if classifier_name == 'LogReg':
        logreg_c = trial.suggest_float("logreg_c", 1e-10, 1e10, log=True)
        classifier_obj = linear_model.LogisticRegression(C=logreg_c)
    else:
        rf_n_estimators = trial.suggest_int("rf_n_estimators", 10, 1000)
        rf_max_depth = trial.suggest_int("rf_max_depth", 2, 32, log=True)
        classifier_obj = ensemble.RandomForestClassifier(
            max_depth=rf_max_depth, n_estimators=rf_n_estimators
        )

    # Step 3: Scoring method:
    score = model_selection.cross_val_score(classifier_obj, X, y, n_jobs=-1, cv=3)
    accuracy = score.mean()
    return accuracy

# Step 4: Running it
study = optuna.create_study(direction="maximize")
study.optimize(objective, n_trials=100)



array([[array([b'Postman, The (1997)', b'Liar Liar (1997)', b'Contact (1997)',
               b'Welcome To Sarajevo (1997)',
               b'I Know What You Did Last Summer (1997)'], dtype=object)      ,
        b'681', array([4., 5., 1., 4., 1.], dtype=float32)],])

array([{'user_id': b'681', 'movie_title': array([b'Postman, The (1997)', b'Liar Liar (1997)', b'Contact (1997)',
              b'Welcome To Sarajevo (1997)',
              b'I Know What You Did Last Summer (1997)'], dtype=object), 'user_rating': array([4., 5., 1., 4., 1.], dtype=float32)},
])

tests/test_ranking.py
import pandas as pd
from sklearn.datasets import load_breast_cancer

import sys
sys.path.insert(0, 'E://neuralnexus')
print(sys.path)

import tensorflow as tf

import neuralnexus as nn

from sklearn.model_selection import train_test_split
from neuralnexus.ranking import RankingModel

from neuralnexus.ranking.utils import get_unique_data, get_unique_query, dataPrep
from neuralnexus.ranking.losses import ListMLELoss

df = pd.read_csv('topic_ranking_data_prep.csv')




unique_data = get_unique_data(df, 'data')

unique_query = get_unique_query(df, 'query')

data = dataPrep(df)

epochs = 10 

cached_train = data.shuffle(100_000).batch(8192).cache()


#list wise sorting 

listwise_model = RankingModel(ListMLELoss(), unique_query, unique_data)
listwise_model.compile(optimizer=tf.keras.optimizers.Adagrad(0.1))

# training
listwise_model.fit(cached_train, epochs=epochs, verbose=False)

# tesing
listwise_model_result = listwise_model.evaluate(cached_train, return_dict=True)
print("NDCG of the ListMLE model: {:.4f}".format(listwise_model_result["ndcg_metric"]))




tests/test_automlclassifier.py
import pandas as pd
from sklearn.datasets import load_breast_cancer

from sklearn.model_selection import train_test_split
from neuralnexus.automl.classifier import Model




def test_classifer():
    d = load_breast_cancer()
    y = d['target']
    X = pd.DataFrame(d['data'],columns = d['feature_names'])

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)


    model = Model()
    model.fit(X_train,y_train)
    assert model.model_status == 200

neuralnexus/ranking/oop.py
import pprint

import numpy as np
import pandas as pd 

import tensorflow as tf
import tensorflow_datasets as tfds

import tensorflow_ranking as tfr
import tensorflow_recommenders as tfrs



class RankingModel(tfrs.Model):

  def __init__(self, loss, unique_query, unique_data):
    super().__init__()
    embedding_dimension = 32

    # Compute embeddings for users.
    self.query_embeddings = tf.keras.Sequential([
      tf.keras.layers.StringLookup(
        vocabulary=unique_query),
      tf.keras.layers.Embedding(len(unique_query) + 2, embedding_dimension)
    ])

    # Compute embeddings for movies.
    self.data_embeddings = tf.keras.Sequential([
      tf.keras.layers.StringLookup(
        vocabulary=unique_data),
      tf.keras.layers.Embedding(len(unique_data) + 2, embedding_dimension)
    ])

    # Compute predictions.
    self.score_model = tf.keras.Sequential([
      # Learn multiple dense layers.
      tf.keras.layers.Dense(256, activation="relu"),
      tf.keras.layers.Dense(64, activation="relu"),
      # Make rating predictions in the final layer.
      tf.keras.layers.Dense(1)
    ])

    self.task = tfrs.tasks.Ranking(
      loss=loss,
      metrics=[
        tfr.keras.metrics.NDCGMetric(name="ndcg_metric"),
        tf.keras.metrics.RootMeanSquaredError()
      ]
    )

  def call(self, features):
    # We first convert the id features into embeddings.
    # User embeddings are a [batch_size, embedding_dim] tensor.
    query_embeddings = self.query_embeddings(features["query"])

    # Movie embeddings are a [batch_size, num_movies_in_list, embedding_dim]
    # tensor.
    data_embeddings = self.data_embeddings(features["data"])

    # We want to concatenate user embeddings with movie emebeddings to pass
    # them into the ranking model. To do so, we need to reshape the user
    # embeddings to match the shape of movie embeddings.
#     print("list_length=========================>", list_length)
    list_length = features["data"].shape[1]
    data_embedding_repeated = tf.repeat(
        tf.expand_dims(query_embeddings, 1), [list_length], axis=1)

    # Once reshaped, we concatenate and pass into the dense layers to generate
    # predictions.
    concatenated_embeddings = tf.concat(
        [data_embedding_repeated, data_embeddings], 2)

    return self.score_model(concatenated_embeddings)

  def compute_loss(self, features, training=False):
    labels = features.pop("revelvence")

    scores = self(features)

    return self.task(
        labels=labels,
        predictions=tf.squeeze(scores, axis=-1),
    )



        

neuralnexus/ranking/utils.py
import numpy as np 
import pandas as pd 

import tensorflow as tf
import tensorflow_datasets as tfds

import tensorflow_ranking as tfr
import tensorflow_recommenders as tfrs



def get_unique_data(df, column):
  """
  `get_unique_data`  take two arguments, dataframe and column. 

  ex:- 

  unique_data = get_unique_data(df, column='data')

  input:-
  [[topic, topic1, topic2], [topic2, topic3, topic1]] 

  Output:-
  [topic, topic1, topic2, topic3]

  get_unique_data will return the unique values of the list of list 
  """
  all_data = []

  for _lst in df[column]:
      all_data = all_data + eval(_lst)
   
  return np.unique(all_data)
    

def get_unique_query(df, column):
  """
  `get_unique_query`  take two arguments, dataframe and column. 

  ex:- 

  unique_query = get_unique_query(df, column='query')

  input:-
  [topic, topic1, topic2, topic2, topic3, topic1] 

  Output:-
  [topic, topic1, topic2, topic3]

  get_unique_query will return the unique values of the list
  """
  return np.unique(list(df[column]))


def dataPrep(df):
  """ 
  dataPrep is convert the dataframe to tf.tensor

  ex:-
  tensor = dataPrep(df)

  condition:- 
  dataframe should have columns in the below given order.

  columns = ['data', 'query', 'revelence']

  data = [topic, topic1, topic2, topic3]

  query = 'name'

  revelence = [0.4, 0.5, 0.9, 0.2]

  """ 
  df.columns = ['data', 'query', 'revelence']

  list_of_data = [eval(i) for i in df['data']]
  list_of_revelence = [eval(i) for i in df['revelence']]

  data = list_of_data
  query = list(df['query'])
  revelence = list_of_revelence

  # convert NumPy arrays to TensorFlow tensors
  data_tensor = tf.convert_to_tensor(data)
  query_tensor = tf.convert_to_tensor(query)
  revelence_tensor = tf.convert_to_tensor(revelence)
  revelence_tensor = tf.cast(revelence_tensor, tf.float32)

  # create a new dictionary with the converted tensors
  my_new_dict = {'data': data_tensor, 
              'query': query_tensor, 
              'revelvence': revelence_tensor}

  # create the TensorSliceDataset
  data = tf.data.Dataset.from_tensor_slices(my_new_dict)
  
  return data

neuralnexus/automl/classifier/oops.py

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer, make_column_selector

from sklearn.preprocessing import StandardScaler, MinMaxScaler, RobustScaler, OneHotEncoder
from sklearn.feature_selection import SelectKBest, f_classif
from sklearn.model_selection import RandomizedSearchCV
from sklearn.pipeline import Pipeline
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, GradientBoostingClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import LinearSVC,SVC
from sklearn.metrics import balanced_accuracy_score



class Model:
  def __init__(self, scoring_function = 'balanced_accuracy', n_iter = 50):
    self.scoring_function = scoring_function
    self.n_iter = n_iter
  
  def fit(self,X,y):
    X_train = X
    y_train = y

    categorical_values = []

    cat_subset = X_train.select_dtypes(include = ['object','category','bool'])

    for i in range(cat_subset.shape[1]):
      categorical_values.append(list(cat_subset.iloc[:,i].dropna().unique()))

    num_pipeline = Pipeline([
                         ('cleaner',SimpleImputer()),
                         ('scaler',StandardScaler())
                         ])

    cat_pipeline = Pipeline([
                        ('cleaner',SimpleImputer(strategy = 'most_frequent')),
                        ('encoder',OneHotEncoder(sparse = False, categories=categorical_values))
    ])


    preprocessor = ColumnTransformer([
      ('numerical', num_pipeline, make_column_selector(dtype_exclude=['object','category','bool'])),
      ('categorical', cat_pipeline, make_column_selector(dtype_include=['object','category','bool']))
    ])

    model_pipeline_steps = []
    model_pipeline_steps.append(('preprocessor',preprocessor))
    model_pipeline_steps.append(('feature_selector',SelectKBest(f_classif,k='all')))
    model_pipeline_steps.append(('estimator',LogisticRegression()))
    model_pipeline = Pipeline(model_pipeline_steps)

    total_features = preprocessor.fit_transform(X_train).shape[1]

    optimization_grid = []

    # Logistic regression
    optimization_grid.append({
        'preprocessor__numerical__scaler':[RobustScaler(),StandardScaler(),MinMaxScaler()],
        'preprocessor__numerical__cleaner__strategy':['mean','median'],
        'feature_selector__k': list(np.arange(1,total_features,5)) + ['all'],
        'estimator':[LogisticRegression()]
    })

    # K-nearest neighbors
    optimization_grid.append({
        'preprocessor__numerical__scaler':[RobustScaler(),StandardScaler(),MinMaxScaler()],
        'preprocessor__numerical__cleaner__strategy':['mean','median'],
        'feature_selector__k': list(np.arange(1,total_features,5)) + ['all'],
        'estimator':[KNeighborsClassifier()],
        'estimator__weights':['uniform','distance'],
        'estimator__n_neighbors':np.arange(1,20,1)
    })

    # Random Forest
    optimization_grid.append({
        'preprocessor__numerical__scaler':[None],
        'preprocessor__numerical__cleaner__strategy':['mean','median'],
        'feature_selector__k': list(np.arange(1,total_features,5)) + ['all'],
        'estimator':[RandomForestClassifier(random_state=0)],
        'estimator__n_estimators':np.arange(5,500,10),
        'estimator__criterion':['gini','entropy']
    })


    # Gradient boosting
    optimization_grid.append({
        'preprocessor__numerical__scaler':[None],
        'preprocessor__numerical__cleaner__strategy':['mean','median'],
        'feature_selector__k': list(np.arange(1,total_features,5)) + ['all'],
        'estimator':[GradientBoostingClassifier(random_state=0)],
        'estimator__n_estimators':np.arange(5,500,10),
        'estimator__learning_rate':np.linspace(0.1,0.9,20),
    })



    # Decision tree
    optimization_grid.append({
        'preprocessor__numerical__scaler':[None],
        'preprocessor__numerical__cleaner__strategy':['mean','median'],
        'feature_selector__k': list(np.arange(1,total_features,5)) + ['all'],
        'estimator':[DecisionTreeClassifier(random_state=0)],
        'estimator__criterion':['gini','entropy']
    })

    # Linear SVM
    optimization_grid.append({
        'preprocessor__numerical__scaler':[RobustScaler(),StandardScaler(),MinMaxScaler()],
        'preprocessor__numerical__cleaner__strategy':['mean','median'],
        'feature_selector__k': list(np.arange(1,total_features,5)) + ['all'],
        'estimator':[LinearSVC(random_state = 0)],
        'estimator__C': np.arange(0.1,1,0.1),
        
    })

    search = RandomizedSearchCV(
      model_pipeline,
      optimization_grid,
      n_iter=self.n_iter,
      scoring = self.scoring_function, 
      n_jobs = -1, 
      random_state = 0, 
      verbose = 3,
      cv = 5
    )

    search.fit(X_train, y_train)
    self.best_estimator_ = search.best_estimator_
    self.best_pipeline = search.best_params_
    self.model_status = 200
    

  
  def predict(self,X,y = None):
    return self.best_estimator_.predict(X)

  def predict_proba(self,X,y = None):
    return self.best_estimator_.predict_proba(X)